### PR TITLE
Implement toggleFit

### LINKED ISSUE #23: Toggle actual size/fit to screen
* Add `ToggleFit` to `Action` enum in `ui.rs`
* Add `actual_size` boolean to program
* Match `z` key to `ToggleFit` in `event_action` function in `ui.rs`
* Match `Action::ToggleFit` to `toggle_fit` function in `run` in `program.rs`
* in `toggle_fit` function flip the `actual_size` boolean
* When changing image in `Next, Prev, Move, First, Last` etc, reset the `actual_size` boolean. It should be true when both images are small than window, and false when any dimension is larger than the corresponding window dimension.
* Add new action to README.

### PR DESCRIPTION
**Pull request by gurgalex**
Fixes #23 

Not sure if the images are exactly centered.

## Details
When each image is switched to, the initial state
of whether to fit to screen or use action size
of image is computed.

Actual size is used if both dimensions of the image are
smaller than the canvas viewport. Otherwise, the image
is fit to screen.

The state can be toggled by pressing `z` or left clicking on the screen.
This change is reset upon changing images.

### CODE REVIEW

**Comment by gurgalex** [comment]
Not sure what rustfmt is finding wrong about on Travis.
Locally, I get nothing.

https://i.imgur.com/QLNHuE7.png

File: src/program/render.rs [hunk 1]
```diff
@@ -54,7 +70,7 @@ impl<'a> Program<'a> {
         {
             return Ok(());
         }
-        let texture = match self
+        let texture: Texture = match self
```
Comment (Davejkane) [inline src/program/render.rs]: What do all these explicit type annotations do?

File: src/program/render.rs [hunk 2]
```diff
@@ -68,11 +84,26 @@ impl<'a> Program<'a> {
...
         self.screen.dirty = false;
         Ok(())
     }
 
+    /// Computes the default state of actual_size for each image
+    pub fn default_actual_size(src_dims: &Rect, dest_dims: &Rect) -> bool {
+        // If any dimension of the src image is bigger than the destination
+        // dimensions, use scaled size.
+        let src_bigger: bool = src_dims.x > dest_dims.x || src_dims.y > dest_dims.y;
+        src_bigger
```
Comment (Davejkane) [inline src/program/render.rs]: can we just compress these two lines to:
`src_dims.x > dest_dims.x || src_dims.y > dest_dims.y`
I think it's clear enough.

**Review by Davejkane — COMMENTED** [review]
I've added a couple of comments looking for some clarification on the explicit type annotations.

The z key just now only toggles between actual size and fitted for images that are larger than the screen. On smaller images, z doesn't do anything. ~I think it should scale up small images as well.~(See below) 

I'm also not sure that we need to copy only a portion of the src to the destination. I think if we just position the rectangle correctly, SDL takes care of the rest efficiently. i.e. the x and y dimensions of the dst rectangle would be negative for large images. Would probably also mean computing the centre of the rectangle would be unnecessary. x would be `(dst_width - src_width) / 2`. Likewise for y `(dst_height - src_height) / 2`. On the other hand, the way you're doing it clearly works.

File: src/program/render.rs (hunk 1 above)
Comment (gurgalex) [reply src/program/render.rs]: Without them, IntelliJ couldn't infer that the type was `&Texture`
And the follwing line `TextureQuery`

https://i.imgur.com/yqhs4rz.png

File: src/program/render.rs (hunk 2 above)
Comment (gurgalex) [reply src/program/render.rs]: Okay

File: src/program/render.rs (hunk 1 above)
Comment (Davejkane) [reply src/program/render.rs]: Hmm. Ok, I don't want to dog on your editor, but vscode seems to figure that out, which is surprising, I thought they both used the RLS underneath. I'm not a big fan of type annotations unless they're required for compilation. But you make a judgement call on this one.

Comment (gurgalex) [reply src/program/render.rs]: Hmm, alright. I'll remove them as the variable names are clear enough about what is expected. Plus, thanks for the editor suggestion.

Comment (Davejkane) [reply src/program/render.rs]: I'm kind of reluctantly using vscode myself. I use emacs for everything else, but for Rust, vscode really is at the front of the pack.

**Review by Davejkane — CHANGES_REQUESTED** [review]
You know what, I just opened a couple of images in my web browser. For ones larger than the screen I can toggle between actual size and fit-to-screen, but for smaller ones, I can't. So I think this implementation is a pretty good default behaviour.  But just like a web browser, can you try implementing this for a left mouse click as well? (remember the README and the help text).

**Comment by gurgalex** [comment]
I was just going to post this as a possible reason for not wanting zoom to fill screen (blurred images for tiny pixel art)

https://i.imgur.com/bgpPxWV.png

**Comment by gurgalex** [comment]
Can always be a config option though for later (user prefs)

**Comment by gurgalex** [comment]
Sure, Left mouse click will toggle between actual and scale to fit on screen.

**Comment by Davejkane** [comment]
Should probably be on mouse button up event (rather than mouse button down). Also based on what my browser does.

**Comment by gurgalex** [comment]
okay, left mouse click is added for `actual_fit` toggle.

**Comment by gurgalex** [comment]
>I'm also not sure that we need to copy only a portion of the src to the destination. I think if we just position the rectangle correctly, SDL takes care of the rest efficiently. i.e. the x and y dimensions of the dst rectangle would be negative for large images. Would probably also mean computing the centre of the rectangle would be unnecessary. x would be (dst_width - src_width) / 2. Likewise for y (dst_height - src_height) / 2. On the other hand, the way you're doing it clearly works.

I'll look into getting the other way working. 

Seems the way you proposed is a more common way.
1st method
https://stackoverflow.com/a/38919425

I need a way that works with offsets, which I believe both can.

https://i.stack.imgur.com/6vCd6.gif

http://lazyfoo.net/tutorials/SDL/30_scrolling/index.php

**Comment by Davejkane** [comment]
@gurgalex, is this ready for merging or are you still planning to change the implementation to not clip the source rectangle?

**Comment by gurgalex** [comment]
It's ready for merging.

**Review by Davejkane — APPROVED** [review]
Cool. Now it's up to you which of your PRs you want to merge first, and which one you want to deal with merge conflicts. I'm sure there won't be much.

**Pull request merged**